File: 2048-kotlin/src/main/kotlin/Game.kt
import kotlinx.browser.document
import kotlinx.browser.window
import org.w3c.dom.*
import org.w3c.dom.events.Event
import org.w3c.dom.events.KeyboardEvent
import kotlin.js.Date
import kotlin.math.min
import kotlin.random.Random

class Game {
    private val canvas: HTMLCanvasElement = document.getElementById("game-canvas") as HTMLCanvasElement
    private val ctx: CanvasRenderingContext2D = canvas.getContext("2d") as CanvasRenderingContext2D
    private var tiles = mutableListOf<Tile>()
    private var score = 0
    private var gameOver = false
    private var startX: Double = 0.0
    private var startY: Double = 0.0
    private var animating = false
    private var animations = mutableListOf<Animation>()
    private var tileSize = 0.0
    private var padding = 0.0

    init {
        handleResize()
        initGame()

        window.addEventListener("keydown", { event -> handleKeyDown(event as KeyboardEvent) })
        window.addEventListener("resize", { handleResize() })
        canvas.addEventListener("touchstart", { event -> handleTouchStart(event) })
        canvas.addEventListener("touchend", { event -> handleTouchEnd(event) })
        document.getElementById("new-game-button")?.addEventListener("click", { initGame() })

        window.requestAnimationFrame { animate() }
    }

    private fun handleResize() {
        canvas.width = canvas.clientWidth
        canvas.height = canvas.clientHeight
        tileSize = min(canvas.width, canvas.height) / 4.0
        padding = tileSize * 0.1
        render()
    }

    private fun initGame() {
        tiles.clear()
        score = 0
        gameOver = false
        document.getElementById("score")?.textContent = "0"
        animations.clear()

        document.querySelector(".game-over")?.let { gameOver ->
            gameOver.parentElement?.removeChild(gameOver)
        }

        addRandomTile()
        addRandomTile()
        render()
    }

    private fun addRandomTile() {
        val emptyCells = mutableListOf<Pair<Int, Int>>()
        for (row in 0..3) {
            for (col in 0..3) {
                if (tiles.none { it.row == row && it.col == col }) {
                    emptyCells.add(row to col)
                }
            }
        }

        if (emptyCells.isNotEmpty()) {
            val (row, col) = emptyCells.random()
            val value = if (Random.nextDouble() < 0.9) 2 else 4
            val tile = Tile(value, row, col)
            tiles.add(tile)
            animations.add(Animation(tile, "scale", 0.0, 1.0, 200.0, Date.now()))
        }
    }

    private fun getTileColor(value: Int): String = when (value) {
        2 -> "#eee4da"
        4 -> "#ede0c8"
        8 -> "#f2b179"
        16 -> "#f59563"
        32 -> "#f67c5f"
        64 -> "#f65e3b"
        128 -> "#edcf72"
        256 -> "#edcc61"
        512 -> "#edc850"
        1024 -> "#edc53f"
        2048 -> "#edc22e"
        else -> "#cdc1b4"
    }

    private fun getTileTextColor(value: Int): String =
        if (value <= 4) "#776e65" else "#f9f6f2"

    private fun move(direction: Direction) {
        if (gameOver || animating) return

        val delta = when (direction) {
            Direction.UP -> -1 to 0
            Direction.DOWN -> 1 to 0
            Direction.LEFT -> 0 to -1
            Direction.RIGHT -> 0 to 1
        }

        var moved = false
        val tilesToRemove = mutableListOf<Tile>()
        animations.clear()

        val sortedTiles = tiles.sortedWith(compareBy { tile ->
            when (direction) {
                Direction.DOWN -> -tile.row
                Direction.UP -> tile.row
                Direction.RIGHT -> -tile.col
                Direction.LEFT -> tile.col
            }
        })

        for (tile in sortedTiles) {
            var newRow = tile.row
            var newCol = tile.col

            while (true) {
                val nextRow = newRow + delta.first
                val nextCol = newCol + delta.second

                if (nextRow !in 0..3 || nextCol !in 0..3) break

                val nextTile = tiles.find { it.row == nextRow && it.col == nextCol && it !in tilesToRemove }
                if (nextTile == null) {
                    newRow = nextRow
                    newCol = nextCol
                    moved = true
                } else if (nextTile.value == tile.value && !tile.merging && !nextTile.merging) {
                    newRow = nextRow
                    newCol = nextCol
                    tile.value *= 2
                    tile.merging = true
                    score += tile.value
                    tilesToRemove.add(nextTile)
                    moved = true
                    break
                } else {
                    break
                }
            }

            if (newRow != tile.row || newCol != tile.col) {
                animations.add(Animation(tile, "row", tile.row.toDouble(), newRow.toDouble(), 150.0, Date.now()))
                animations.add(Animation(tile, "col", tile.col.toDouble(), newCol.toDouble(), 150.0, Date.now()))
                if (tile.merging) {
                    animations.add(Animation(tile, "scale", 1.0, 1.2, 100.0, Date.now()))
                    animations.add(Animation(tile, "scale", 1.2, 1.0, 100.0, Date.now() + 100))
                }
                tile.row = newRow
                tile.col = newCol
            }
        }

        if (moved) {
            animating = true
            document.getElementById("score")?.textContent = score.toString()

            window.setTimeout({
                tiles.removeAll(tilesToRemove)
                tiles.forEach { it.merging = false }
                addRandomTile()
                animating = false
                checkGameOver()
            }, 150)
        }
    }

    private fun checkGameOver() {
        if (tiles.size < 16) return

        for (row in 0..3) {
            for (col in 0..3) {
                val tile = tiles.find { it.row == row && it.col == col } ?: continue
                val value = tile.value

                // Check adjacent cells
                val adjacentTiles = listOfNotNull(
                    if (row > 0) tiles.find { it.row == row - 1 && it.col == col } else null,
                    if (row < 3) tiles.find { it.row == row + 1 && it.col == col } else null,
                    if (col > 0) tiles.find { it.row == row && it.col == col - 1 } else null,
                    if (col < 3) tiles.find { it.row == row && it.col == col + 1 } else null
                )

                if (adjacentTiles.any { it.value == value }) return
            }
        }

        gameOver = true
        showGameOver()
    }

    private fun showGameOver() {
        val gameOverDiv = document.createElement("div") as HTMLDivElement
        gameOverDiv.className = "game-over"
        gameOverDiv.innerHTML = """
            <h2>Game Over!</h2>
            <button id="new-game-button">Try Again</button>
        """.trimIndent()
        document.querySelector(".game-container")?.appendChild(gameOverDiv)
        gameOverDiv.querySelector("button")?.addEventListener("click", { initGame() })
    }

    private fun handleKeyDown(event: KeyboardEvent) {
        when (event.key) {
            "ArrowUp" -> {
                event.preventDefault()
                move(Direction.UP)
            }
            "ArrowDown" -> {
                event.preventDefault()
                move(Direction.DOWN)
            }
            "ArrowLeft" -> {
                event.preventDefault()
                move(Direction.LEFT)
            }
            "ArrowRight" -> {
                event.preventDefault()
                move(Direction.RIGHT)
            }
        }
    }

    private fun handleTouchStart(event: Event) {
        val touchEvent = event.asDynamic()
        startX = touchEvent.touches[0].clientX as Double
        startY = touchEvent.touches[0].clientY as Double
    }

    private fun handleTouchEnd(event: Event) {
        val touchEvent = event.asDynamic()
        val deltaX = (touchEvent.changedTouches[0].clientX as Double) - startX
        val deltaY = (touchEvent.changedTouches[0].clientY as Double) - startY
        val minSwipeDistance = 50

        if (kotlin.math.abs(deltaX) > kotlin.math.abs(deltaY)) {
            if (kotlin.math.abs(deltaX) > minSwipeDistance) {
                move(if (deltaX > 0) Direction.RIGHT else Direction.LEFT)
            }
        } else {
            if (kotlin.math.abs(deltaY) > minSwipeDistance) {
                move(if (deltaY > 0) Direction.DOWN else Direction.UP)
            }
        }
    }

    private fun animate(timestamp: Double = 0.0) {
        render()
        window.requestAnimationFrame { animate(it) }
    }

    private fun render() {
        val now = Date.now()
        ctx.clearRect(0.0, 0.0, canvas.width.toDouble(), canvas.height.toDouble())

        // Draw grid
        ctx.fillStyle = "#cdc1b4"
        for (row in 0..3) {
            for (col in 0..3) {
                val x = col * tileSize + padding
                val y = row * tileSize + padding
                val size = tileSize - padding * 2
                ctx.fillRect(x, y, size, size)
            }
        }

        // Update animations
        animations.removeAll { updateAnimation(it, now) }

        // Draw tiles
        for (tile in tiles) {
            val scale = tile.scale
            val x = tile.col * tileSize + padding
            val y = tile.row * tileSize + padding
            val size = (tileSize - padding * 2) * scale
            val offsetX = (tileSize - size) / 2
            val offsetY = (tileSize - size) / 2

            // Draw tile background
            ctx.fillStyle = getTileColor(tile.value)
            ctx.fillRect(x + offsetX, y + offsetY, size, size)

            // Draw tile text
            ctx.fillStyle = getTileTextColor(tile.value)
            ctx.font = "bold ${size * 0.4}px Arial"
            ctx.textAlign = CanvasTextAlign.CENTER
            ctx.textBaseline = CanvasTextBaseline.MIDDLE
            ctx.fillText(
                tile.value.toString(),
                x + offsetX + size / 2,
                y + offsetY + size / 2
            )
        }
    }

    private fun updateAnimation(animation: Animation, now: Double): Boolean {
        val elapsed = now - animation.startTime
        if (elapsed >= animation.duration) {
            when (animation.property) {
                "row" -> animation.tile.row = animation.end.toInt()
                "col" -> animation.tile.col = animation.end.toInt()
                "scale" -> animation.tile.scale = animation.end
            }
            return true
        }
        val progress = elapsed / animation.duration
        val value = animation.start + (animation.end - animation.start) * progress
        when (animation.property) {
            "row" -> animation.tile.row = value.toInt()
            "col" -> animation.tile.col = value.toInt()
            "scale" -> animation.tile.scale = value
        }
        return false
    }

    private enum class Direction {
        UP, DOWN, LEFT, RIGHT
    }

    private data class Tile(
        var value: Int,
        var row: Int,
        var col: Int,
        var scale: Double = 1.0,
        var merging: Boolean = false
    )

    private data class Animation(
        val tile: Tile,
        val property: String,
        val start: Double,
        val end: Double,
        val duration: Double,
        val startTime: Double
    )
}

fun main() {
    Game()
}

File: 2048-pyscript/game.py
from js import document, window, requestAnimationFrame, performance
import random
import math
from pyodide.ffi import create_proxy

class Game:
    def __init__(self):
        self.canvas = document.getElementById('game-canvas')
        self.ctx = self.canvas.getContext('2d')
        self.tiles = []
        self.score = 0
        self.game_over = False
        self.touch_start = None
        self.animating = False
        self.animations = []

        # Create proxies for event handlers
        self.handle_key_down_proxy = create_proxy(self.handle_key_down)
        self.handle_resize_proxy = create_proxy(self.handle_resize)
        self.handle_touch_start_proxy = create_proxy(self.handle_touch_start)
        self.handle_touch_end_proxy = create_proxy(self.handle_touch_end)
        self.init_game_proxy = create_proxy(self.init_game)
        self.animate_proxy = create_proxy(self.animate)

        # Set canvas size
        self.handle_resize()

        # Initialize game
        self.init_game()

        # Event listeners
        window.addEventListener('keydown', self.handle_key_down_proxy)
        window.addEventListener('resize', self.handle_resize_proxy)
        self.canvas.addEventListener('touchstart', self.handle_touch_start_proxy)
        self.canvas.addEventListener('touchend', self.handle_touch_end_proxy)
        document.getElementById('new-game-button').addEventListener('click', self.init_game_proxy)

        # Start animation loop
        requestAnimationFrame(self.animate_proxy)

    def handle_resize(self, event=None):
        width = self.canvas.clientWidth
        height = self.canvas.clientHeight
        
        self.canvas.width = width
        self.canvas.height = height
        self.tile_size = min(width, height) / 4
        self.padding = self.tile_size * 0.1

        self.render()

    def init_game(self):
        self.tiles = []
        self.score = 0
        self.game_over = False
        document.getElementById('score').textContent = '0'
        self.animations = []

        # Remove game over screen if exists
        game_over = document.querySelector('.game-over')
        if game_over:
            game_over.parentElement.removeChild(game_over)

        # Add initial tiles
        self.add_random_tile()
        self.add_random_tile()
        self.render()

    def add_random_tile(self):
        empty_cells = []
        for row in range(4):
            for col in range(4):
                if not any(tile['row'] == row and tile['col'] == col for tile in self.tiles):
                    empty_cells.append({'row': row, 'col': col})

        if empty_cells:
            cell = random.choice(empty_cells)
            value = 2 if random.random() < 0.9 else 4
            tile = {
                'value': value,
                'row': cell['row'],
                'col': cell['col'],
                'scale': 0,
                'merging': False
            }
            self.tiles.append(tile)
            self.animations.append({
                'tile': tile,
                'property': 'scale',
                'start': 0,
                'end': 1,
                'duration': 200,
                'start_time': performance.now()
            })

    def get_tile_color(self, value):
        colors = {
            2: '#eee4da',
            4: '#ede0c8',
            8: '#f2b179',
            16: '#f59563',
            32: '#f67c5f',
            64: '#f65e3b',
            128: '#edcf72',
            256: '#edcc61',
            512: '#edc850',
            1024: '#edc53f',
            2048: '#edc22e'
        }
        return colors.get(value, '#cdc1b4')

    def get_tile_text_color(self, value):
        return '#776e65' if value <= 4 else '#f9f6f2'

    def move(self, direction):
        if self.game_over or self.animating:
            return

        direction_map = {
            'up': {'row_delta': -1, 'col_delta': 0},
            'down': {'row_delta': 1, 'col_delta': 0},
            'left': {'row_delta': 0, 'col_delta': -1},
            'right': {'row_delta': 0, 'col_delta': 1}
        }

        delta = direction_map[direction]
        moved = False
        tiles_to_remove = []
        self.animations = []

        # Sort tiles based on direction
        sorted_tiles = sorted(self.tiles, key=lambda t: (
            -t['row'] if direction == 'down' else t['row'] if direction == 'up' else
            -t['col'] if direction == 'right' else t['col']
        ))

        # Move tiles
        for tile in sorted_tiles:
            row, col = tile['row'], tile['col']
            new_row, new_col = row, col

            while True:
                next_row = new_row + delta['row_delta']
                next_col = new_col + delta['col_delta']

                if next_row < 0 or next_row > 3 or next_col < 0 or next_col > 3:
                    break

                next_tile = next((t for t in self.tiles if t['row'] == next_row and t['col'] == next_col and t not in tiles_to_remove), None)
                if not next_tile:
                    new_row = next_row
                    new_col = next_col
                    moved = True
                elif next_tile['value'] == tile['value'] and not tile['merging'] and not next_tile['merging']:
                    new_row = next_row
                    new_col = next_col
                    tile['value'] *= 2
                    tile['merging'] = True
                    self.score += tile['value']
                    tiles_to_remove.append(next_tile)
                    moved = True
                    break
                else:
                    break

            if new_row != row or new_col != col:
                self.animations.extend([
                    {
                        'tile': tile,
                        'property': 'row',
                        'start': row,
                        'end': new_row,
                        'duration': 150,
                        'start_time': performance.now()
                    },
                    {
                        'tile': tile,
                        'property': 'col',
                        'start': col,
                        'end': new_col,
                        'duration': 150,
                        'start_time': performance.now()
                    }
                ])
                if tile['merging']:
                    self.animations.extend([
                        {
                            'tile': tile,
                            'property': 'scale',
                            'start': 1,
                            'end': 1.2,
                            'duration': 100,
                            'start_time': performance.now()
                        },
                        {
                            'tile': tile,
                            'property': 'scale',
                            'start': 1.2,
                            'end': 1,
                            'duration': 100,
                            'start_time': performance.now() + 100
                        }
                    ])
                tile['row'] = new_row
                tile['col'] = new_col

        if moved:
            self.animating = True
            document.getElementById('score').textContent = str(self.score)

            def after_move():
                self.tiles = [t for t in self.tiles if t not in tiles_to_remove]
                for tile in self.tiles:
                    tile['merging'] = False
                self.add_random_tile()
                self.animating = False
                self.check_game_over()

            window.setTimeout(create_proxy(after_move), 150)

    def check_game_over(self):
        if len(self.tiles) < 16:
            return

        for row in range(4):
            for col in range(4):
                tile = next((t for t in self.tiles if t['row'] == row and t['col'] == col), None)
                if tile:
                    value = tile['value']
                    # Check adjacent cells
                    adjacent_tiles = [
                        next((t for t in self.tiles if t['row'] == row - 1 and t['col'] == col), None) if row > 0 else None,
                        next((t for t in self.tiles if t['row'] == row + 1 and t['col'] == col), None) if row < 3 else None,
                        next((t for t in self.tiles if t['row'] == row and t['col'] == col - 1), None) if col > 0 else None,
                        next((t for t in self.tiles if t['row'] == row and t['col'] == col + 1), None) if col < 3 else None
                    ]
                    
                    for adj_tile in adjacent_tiles:
                        if adj_tile is not None and adj_tile.get('value') == value:
                            return

        self.game_over = True
        self.show_game_over()

    def show_game_over(self):
        game_over = document.createElement('div')
        game_over.className = 'game-over'
        game_over.innerHTML = '''
            <h2>Game Over!</h2>
            <button id="new-game-button">Try Again</button>
        '''
        document.querySelector('.game-container').appendChild(game_over)
        game_over.querySelector('button').addEventListener('click', create_proxy(self.init_game))

    def handle_key_down(self, event):
        key_map = {
            'ArrowUp': 'up',
            'ArrowDown': 'down',
            'ArrowLeft': 'left',
            'ArrowRight': 'right'
        }
        if event.key in key_map:
            event.preventDefault()
            self.move(key_map[event.key])

    def handle_touch_start(self, event):
        touch = event.touches[0]
        self.touch_start = {'x': touch.clientX, 'y': touch.clientY}

    def handle_touch_end(self, event):
        if not self.touch_start:
            return

        touch = event.changedTouches[0]
        delta_x = touch.clientX - self.touch_start['x']
        delta_y = touch.clientY - self.touch_start['y']
        min_swipe_distance = 50

        if abs(delta_x) > abs(delta_y):
            if abs(delta_x) > min_swipe_distance:
                self.move('right' if delta_x > 0 else 'left')
        else:
            if abs(delta_y) > min_swipe_distance:
                self.move('down' if delta_y > 0 else 'up')

        self.touch_start = None

    def animate(self, timestamp=None):
        self.render()
        requestAnimationFrame(self.animate_proxy)

    def render(self):
        now = performance.now()
        self.ctx.clearRect(0, 0, self.canvas.width, self.canvas.height)

        # Draw grid
        self.ctx.fillStyle = '#cdc1b4'
        for row in range(4):
            for col in range(4):
                x = col * self.tile_size + self.padding
                y = row * self.tile_size + self.padding
                size = self.tile_size - self.padding * 2
                self.ctx.fillRect(x, y, size, size)

        # Update animations
        self.animations = [anim for anim in self.animations if self.update_animation(anim, now)]

        # Draw tiles
        for tile in self.tiles:
            scale = tile.get('scale', 1)
            x = tile['col'] * self.tile_size + self.padding
            y = tile['row'] * self.tile_size + self.padding
            size = (self.tile_size - self.padding * 2) * scale
            offset_x = (self.tile_size - size) / 2
            offset_y = (self.tile_size - size) / 2

            # Draw tile background
            self.ctx.fillStyle = self.get_tile_color(tile['value'])
            self.ctx.fillRect(x + offset_x, y + offset_y, size, size)

            # Draw tile text
            self.ctx.fillStyle = self.get_tile_text_color(tile['value'])
            self.ctx.font = f'bold {size * 0.4}px Arial'
            self.ctx.textAlign = 'center'
            self.ctx.textBaseline = 'middle'
            self.ctx.fillText(
                str(tile['value']),
                x + offset_x + size / 2,
                y + offset_y + size / 2
            )

    def update_animation(self, animation, now):
        elapsed = now - animation['start_time']
        if elapsed >= animation['duration']:
            animation['tile'][animation['property']] = animation['end']
            return False
        progress = elapsed / animation['duration']
        animation['tile'][animation['property']] = animation['start'] + (animation['end'] - animation['start']) * progress
        return True

# Initialize game
game = Game()